### PR TITLE
Create OBC for registered device

### PR DESCRIPTION
**Pull request by masayag**
Upon registration of edge device, the k4e controller will create an OBC
and the required data to access the bucket will be set as part of device
configuration in response to get-data-message requests.

Signed-off-by: Moti Asayag <masayag@redhat.com>

### CODE REVIEW

File: internal/storage/storage.go [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+	if !exist {
+		return nil, err
+	}
+	conf.AwsAccessKeyID = base64.StdEncoding.EncodeToString(awsAccessKeyID)
+	awsSecretAccessKey, exist := secret.Data["AWS_SECRET_ACCESS_KEY"]
+	if !exist {
+		return nil, err
+	}
+	conf.AwsSecretAccessKey = base64.StdEncoding.EncodeToString(awsSecretAccessKey)
+	return conf, nil
```
Comment (masayag) [inline internal/storage/storage.go:79]: the returned object should look like:
```
"storageConf": {
  "aws_access_key_id": "MTU3SmRSVjNyV0xObUhTUmZ6QTg=",
  "aws_secret_access_key": "K1RBdWVmbWZZZk95eFMvRXNrRlVzaTNJUjhqaHJ2bU9kdHNzYThkRA==",
  "bucket_host": "s3.openshift-storage.svc",
  "bucket_name": "07b68c4e-41b0-4cb9-87ef-17-cf93bc4e-385c-4545-be64-ac3f8e07427c",
  "bucket_port": 443
}
```

**Review by pkliczewski — APPROVED** [review]

File: controllers/edgedevice_controller.go [hunk 2]
```diff
@@ -61,6 +68,28 @@ func (r *EdgeDeviceReconciler) Reconcile(ctx context.Context, req ctrl.Request)
...
+		if err != nil {
+			if errors.IsNotFound(err) {
+				logger.Info("Failed to find an existing OBC for the device. Creating new OBC", "edgeDevice", edgeDevice)
+				obc, err = r.Claimer.CreateClaim(ctx, edgeDevice)
+				if err != nil {
+					logger.Error(err, "Cannot create object bucket claim for the device", "EdgeDevice Name", edgeDevice.Name, "EdgeDevice Namespace", edgeDevice.Namespace)
+					return ctrl.Result{Requeue: true}, err
+				}
+			} else {
+				logger.Error(err, "Failed to get OBC for the device", "edgeDevice", edgeDevice)
```
Comment (jakub-dzon) [inline controllers/edgedevice_controller.go:83]: Shouldn't `ctrl.Result{Requeue: true}, err` be returned in this case? `obc` would be `nil` and as a result line 87 would panic.

**Review by jakub-dzon — CHANGES_REQUESTED** [review]

File: internal/storage/storage.go (hunk 1 above)
Comment (jakub-dzon) [reply internal/storage/storage.go:79]: The agent/device needs the `bucket_host`to be a routable fqdn in cluster's domain.
_Accessing a bucket from outside the cluster_ section of https://blog.oddbit.com/post/2021-02-10-object-storage-with-openshift lists a script, that prepares bucket_host that can be used by a device.

File: controllers/edgedevice_controller.go (hunk 2 above)
Comment (masayag) [reply controllers/edgedevice_controller.go:83]: Fixed.

File: internal/storage/storage.go (hunk 1 above)
Comment (masayag) [reply internal/storage/storage.go:82]: You're right, I'll update the returned configuration to return:
```
{
  "aws_access_key_id": "enZvYWNWeU9UcUp4d29vRkgxanY=",
  "aws_secret_access_key": "dW5idWdjTHlOeXlVQ21HZEF4aU8zOEwwWU5TZzdrNm8rb01tSmZXNg==",
  "bucket_host": "s3-openshift-storage.apps.test-infra-cluster-f13cd1b3.redhat.com",
  "bucket_name": "07b68c4e-41b0-4cb9-87ef-17-97e6547c-5da4-4e96-b001-88302ceb9c32",
  "bucket_port": 443
}
```

Comment (masayag) [reply internal/storage/storage.go:96]: With this values, it is possible to connect to the s3 endpoint successfully.
I also added the `AWS_CA_BUNDLE` property to store the required certificate for the SSL endpoint (base64 encoded, as the access-key and the secret-access-key).
```
AWS_SECRET_ACCESS_KEY=unbugcLyNyyUCmGdAxiO38L0YNSg7k6o+oMmJfW6
AWS_ACCESS_KEY_ID=zvoacVyOTqJxwooFH1jv
AWS_CA_BUNDLE=/tmp/s3.crt
BUCKET_PORT=443
BUCKET_NAME=07b68c4e-41b0-4cb9-87ef-17-97e6547c-5da4-4e96-b001-88302ceb9c32
BUCKET_HOST=s3-openshift-storage.apps.test-infra-cluster-f13cd1b3.redhat.com
→ aws s3  --endpoint https://$BUCKET_HOST ls --human-readable --summarize
2021-09-05 15:12:22 07b68c4e-41b0-4cb9-87ef-17-97e6547c-5da4-4e96-b001-88302ceb9c32

Total Objects: 0
   Total Size: 0 Bytes
```

**Review by jakub-dzon — APPROVED** [review]
Looks great!

**Pull request merged**